Unpacking starred expressions in tuples

(migrated from [a comment](https://github.com/python/mypy/issues/1178#issuecomment-545327224) in #1178 at @ethanhs's request)

There seems to an issue with unpacking things when they are part of a tuple on the right side of assignment. #1178 is related, but specifically deals with variable-length tuples (and using `len()` to distinguish them). The current issue occurs even for fixed-length and type-annotated tuples.

```python
# file: ex.py
from typing import Tuple
points = (1, 2)  # type: Tuple[int, int]
x, y, z = *points, 0
```

This results in incorrect inference about the tuple size (but Python does not complain, only mypy):

```console
$ mypy ex.py 
ex.py:4: error: Need more than 2 values to unpack (3 expected)
Found 1 error in 1 file (checked 1 source file)
```

A workaround is to use cast(), as @JukkaL [suggested](https://github.com/python/mypy/issues/1178#issuecomment-176185607):

```python
x, y, z = cast(Tuple[int, int, int], (*points, 0))
```

And here's my versions of Python and mypy:

```console
$ mypy --version
mypy 0.740
$ python --version
Python 3.7.3
```

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -2483,36 +2483,47 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
             # control in cases like: a, b = [int, str] where rhs would get
             # type List[object]
             rvalues = []  # type: List[Expression]
-            lhs_len = len(lvalues)
-            idx_of_iterable = []
-            item_type_of_iterable = []  # type: List['mypy.types.Type']
-            idx = 0
-            for rval in rvalue.items:
+            iterable_type = None  # type: Optional[Type]
+            last_idx = None  # type: Optional[int]
+            for idx_rval, rval in enumerate(rvalue.items):
                 if isinstance(rval, StarExpr):
                     typs = get_proper_type(self.expr_checker.visit_star_expr(rval).type)
                     if isinstance(typs, TupleType):
                         rvalues.extend([TempNode(typ) for typ in typs.items])
-                        lhs_len -= len(typs.items)
-                        idx += len(typs.items)
                     elif self.type_is_iterable(typs) and isinstance(typs, Instance):
-                        item_type_of_iterable.append(self.iterable_item_type(typs))
-                        idx_of_iterable.append(idx)
+                        if (iterable_type is not None
+                                and iterable_type != self.iterable_item_type(typs)):
+                            self.fail("Contiguous iterable with same type expected", context)
+                        else:
+                            if last_idx is None or last_idx + 1 == idx_rval:
+                                rvalues.append(rval)
+                                last_idx = idx_rval
+                                iterable_type = self.iterable_item_type(typs)
+                            else:
+                                self.fail("Contiguous iterable with same type expected", context)
                     else:
-                        self.fail("StarExpr should not be a '{}'".format(typs), context)
+                        self.fail("Invalid type '{}' for *expr (iterable expected)".format(typs),
+                             context)
                 else:
                     rvalues.append(rval)
-                    lhs_len -= 1
-                    idx += 1
-            num_every_iterable = 0
-            num_last_iterable = 0
-            if len(idx_of_iterable):
-                num_every_iterable = int(lhs_len / len(idx_of_iterable))
-                num_last_iterable = lhs_len - (len(idx_of_iterable) - 1) * int(num_every_iterable)
-            for i, (idx, item_type) in enumerate(zip(idx_of_iterable, item_type_of_iterable)):
-                if i == (len(idx_of_iterable) - 1):
-                    rvalues[idx:idx] = [TempNode(item_type) for _ in range(num_last_iterable)]
-                else:
-                    rvalues[idx:idx] = [TempNode(item_type) for _ in range(num_every_iterable)]
+            iterable_start = None  # type: Optional[int]
+            iterable_end = None  # type: Optional[int]
+            for i, rval in enumerate(rvalues):
+                if isinstance(rval, StarExpr):
+                    typs = get_proper_type(self.expr_checker.visit_star_expr(rval).type)
+                    if self.type_is_iterable(typs) and isinstance(typs, Instance):
+                        if iterable_start is None:
+                            iterable_start = i
+                        iterable_end = i
+            if (iterable_start is not None
+                    and iterable_end is not None
+                    and iterable_type is not None):
+                iterable_num = iterable_end - iterable_start + 1
+                rvalue_needed = len(lvalues) - (len(rvalues) - iterable_num)
+                if rvalue_needed > 0:
+                    rvalues = rvalues[0: iterable_start] + [TempNode(iterable_type)
+                        for i in range(rvalue_needed)] + rvalues[iterable_end + 1:]
+
             if self.check_rvalue_count_in_assignment(lvalues, len(rvalues), context):
                 star_index = next((i for i, lv in enumerate(lvalues) if
                                    isinstance(lv, StarExpr)), len(lvalues))

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -2483,8 +2483,16 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
             # control in cases like: a, b = [int, str] where rhs would get
             # type List[object]
 
-            rvalues = rvalue.items
-
+            rvalues = []  # type: List[Expression]
+            for rval in rvalue.items:
+                if isinstance(rval, StarExpr):
+                    typs = get_proper_type(self.expr_checker.visit_star_expr(rval).type)
+                    if isinstance(typs, TupleType):
+                        rvalues.extend([TempNode(typ) for typ in typs.items])
+                    else:
+                        rvalues.append(TempNode(typs))
+                else:
+                    rvalues.append(rval)
             if self.check_rvalue_count_in_assignment(lvalues, len(rvalues), context):
                 star_index = next((i for i, lv in enumerate(lvalues) if
                                    isinstance(lv, StarExpr)), len(lvalues))

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -2482,17 +2482,37 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
             # using the type of rhs, because this allowed more fine grained
             # control in cases like: a, b = [int, str] where rhs would get
             # type List[object]
-
             rvalues = []  # type: List[Expression]
+            lhs_len = len(lvalues)
+            idx_of_iterable = []
+            item_type_of_iterable = []  # type: List['mypy.types.Type']
+            idx = 0
             for rval in rvalue.items:
                 if isinstance(rval, StarExpr):
                     typs = get_proper_type(self.expr_checker.visit_star_expr(rval).type)
                     if isinstance(typs, TupleType):
                         rvalues.extend([TempNode(typ) for typ in typs.items])
+                        lhs_len -= len(typs.items)
+                        idx += len(typs.items)
+                    elif self.type_is_iterable(typs) and isinstance(typs, Instance):
+                        item_type_of_iterable.append(self.iterable_item_type(typs))
+                        idx_of_iterable.append(idx)
                     else:
-                        rvalues.append(TempNode(typs))
+                        self.fail("StarExpr should not be a '{}'".format(typs), context)
                 else:
                     rvalues.append(rval)
+                    lhs_len -= 1
+                    idx += 1
+            num_every_iterable = 0
+            num_last_iterable = 0
+            if len(idx_of_iterable):
+                num_every_iterable = int(lhs_len / len(idx_of_iterable))
+                num_last_iterable = lhs_len - (len(idx_of_iterable) - 1) * int(num_every_iterable)
+            for i, (idx, item_type) in enumerate(zip(idx_of_iterable, item_type_of_iterable)):
+                if i == (len(idx_of_iterable) - 1):
+                    rvalues[idx:idx] = [TempNode(item_type) for _ in range(num_last_iterable)]
+                else:
+                    rvalues[idx:idx] = [TempNode(item_type) for _ in range(num_every_iterable)]
             if self.check_rvalue_count_in_assignment(lvalues, len(rvalues), context):
                 star_index = next((i for i, lv in enumerate(lvalues) if
                                    isinstance(lv, StarExpr)), len(lvalues))

--- a/test-data/unit/check-tuples.test
+++ b/test-data/unit/check-tuples.test
@@ -1445,4 +1445,16 @@ reveal_type(y3) # N: Revealed type is 'builtins.int*'
 reveal_type(y4) # N: Revealed type is 'builtins.int*'
 reveal_type(z3) # N: Revealed type is 'builtins.str'
 
+x5, x6, y5, y6, z4 = *points2, *points2, "test"
+
+reveal_type(x5) # N: Revealed type is 'builtins.int*'
+reveal_type(x6) # N: Revealed type is 'builtins.int*'
+reveal_type(y5) # N: Revealed type is 'builtins.int*'
+reveal_type(y6) # N: Revealed type is 'builtins.int*'
+reveal_type(z4) # N: Revealed type is 'builtins.str'
+
+points3 = ["test1", "test2"]
+x7, x8, y7, y8 = *points2, *points3 # E: Contiguous iterable with same type expected
+
+x9, y9, x10, y10, z5 = *points2, 1, *points2 # E: Contiguous iterable with same type expected
 [builtins fixtures/tuple.pyi]

--- a/test-data/unit/check-tuples.test
+++ b/test-data/unit/check-tuples.test
@@ -1421,3 +1421,10 @@ t6: Tuple[int, int, int, int, int, int, int, int, int, int, int, int] = (1, 2, 3
                                                                     # E: Incompatible types in assignment (expression has type Tuple[int, int, ... <15 more items>], variable has type Tuple[int, int, ... <10 more items>])
 
 [builtins fixtures/tuple.pyi]
+
+[case testTupleWithStarExpr]
+from typing import Tuple
+points = (1, 2)  # type: Tuple[int, int]
+x, y, z = *points, 0
+
+[builtins fixtures/tuple.pyi]

--- a/test-data/unit/check-tuples.test
+++ b/test-data/unit/check-tuples.test
@@ -1423,8 +1423,26 @@ t6: Tuple[int, int, int, int, int, int, int, int, int, int, int, int] = (1, 2, 3
 [builtins fixtures/tuple.pyi]
 
 [case testTupleWithStarExpr]
-from typing import Tuple
-points = (1, 2)  # type: Tuple[int, int]
+from typing import Tuple, List
+points = (1, "test")  # type: Tuple[int, str]
 x, y, z = *points, 0
+reveal_type(x) # N: Revealed type is 'builtins.int'
+reveal_type(y) # N: Revealed type is 'builtins.str'
+reveal_type(z) # N: Revealed type is 'builtins.int'
+
+points2 = [1,2]
+x2, y2, z2= *points2, "test"
+
+reveal_type(x2) # N: Revealed type is 'builtins.int*'
+reveal_type(y2) # N: Revealed type is 'builtins.int*'
+reveal_type(z2) # N: Revealed type is 'builtins.str'
+
+x3, x4, y3, y4, z3 = *points, *points2, "test"
+
+reveal_type(x3) # N: Revealed type is 'builtins.int'
+reveal_type(x4) # N: Revealed type is 'builtins.str'
+reveal_type(y3) # N: Revealed type is 'builtins.int*'
+reveal_type(y4) # N: Revealed type is 'builtins.int*'
+reveal_type(z3) # N: Revealed type is 'builtins.str'
 
 [builtins fixtures/tuple.pyi]

